x/ccv/provider/keeper/distribution.go

```diff
@@ -1,10 +1,141 @@
 package keeper
 
 import (
+	"errors"
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/interchain-security/x/ccv/provider/types"
+	ccv "github.com/cosmos/interchain-security/x/ccv/types"
 )
 
 func (k Keeper) GetFeeCollectorAddressStr(ctx sdk.Context) string {
 	return k.accountKeeper.GetModuleAccount(
 		ctx, k.feeCollectorName).GetAddress().String()
 }
+
+func (k Keeper) GetProviderDistributionAddressStr(ctx sdk.Context) string {
+	return k.accountKeeper.GetModuleAccount(
+		ctx, types.DistributionAccount).GetAddress().String()
+}
+
+func (k Keeper) HandleProviderPoolWeights(ctx sdk.Context, weights ccv.ProviderPoolWeights) error {
+	if weights.GetTokens().IsZero() {
+		return errors.New("Tokens should not be zero ")
+	}
+	if len(weights.Addresses) == 0 {
+		return errors.New("Addresses should not be empty ")
+	}
+	if len(weights.Addresses) != len(weights.Weights) {
+		return errors.New("Addresses size does not equal to weights size ")
+	}
+	totalWeights := sdk.ZeroInt()
+	for _, weight := range weights.Weights {
+		totalWeights = totalWeights.Add(weight)
+	}
+	if !totalWeights.Equal(weights.TotalWeight) {
+		return errors.New("TotalWeight does not equal to the sum of weights")
+	}
+
+	k.AddPendingProviderPoolWeights(ctx, &weights)
+	return nil
+}
+
+func (k Keeper) Distribute(ctx sdk.Context) {
+	address := k.accountKeeper.GetModuleAccount(
+		ctx, types.DistributionAccount).GetAddress()
+	balances := k.bankKeeper.GetAllBalances(ctx, address)
+	if balances.IsZero() {
+		return
+	}
+
+	pending := k.GetPendingProviderPoolWeights(ctx)
```

Comment:
"GetPendingProviderPoolWeights" & "SetPendingProviderPoolWeights" need to be atomic, incase "k.SetPendingProviderPoolWeights(ctx, ccv.PendingPPWs{})" overwrite HandleProviderPoolWeights udpates